platform/iphone/app_delegate.mm

```diff
@@ -603,111 +125,56 @@ - (void)applicationDidReceiveMemoryWarning:(UIApplication *)application {
+	NSString *documents = paths.firstObject;
+	int err = iphone_main(vm.width, vm.height, gargc, gargv, String::utf8([documents UTF8String]));
 	if (err != 0) {
 		// bail, things did not go very well for us, should probably output a message on screen with our error code...
 		exit(0);
-		return FALSE;
+		return NO;
 	};
 
-	view_controller = [[ViewController alloc] init];
-	view_controller.view = glView;
-	window.rootViewController = view_controller;
-
 	_set_keep_screen_on(bool(GLOBAL_DEF("display/window/energy_saving/keep_screen_on", true)) ? YES : NO);
-	glView.useCADisplayLink =
-			bool(GLOBAL_DEF("display.iOS/use_cadisplaylink", true)) ? YES : NO;
-	printf("cadisaplylink: %d", glView.useCADisplayLink);
-	glView.animationInterval = 1.0 / kRenderingFrequency;
-	[glView startAnimation];
 
-	// Show the window
-	[window makeKeyAndVisible];
-
-	// Configure and start accelerometer
-	if (!motionInitialised) {
-		motionManager = [[CMMotionManager alloc] init];
-		if (motionManager.deviceMotionAvailable) {
-			motionManager.deviceMotionUpdateInterval = 1.0 / 70.0;
-			[motionManager startDeviceMotionUpdatesUsingReferenceFrame:
-								   CMAttitudeReferenceFrameXMagneticNorthZVertical];
-			motionInitialised = YES;
-		};
-	};
+	// Create our engine view controller
+	GodotGameViewController *rootVC = [[GodotGameViewController alloc] init];
+	self.window.rootViewController = rootVC;
 
-	[self initGameControllers];
+	GLView *glView = rootVC.glView;
+	glView.delegate = self;
+	[glView startAnimation];
 
 	[[NSNotificationCenter defaultCenter]
 			addObserver:self
 			   selector:@selector(onAudioInterruption:)
 				   name:AVAudioSessionInterruptionNotification
 				 object:[AVAudioSession sharedInstance]];
 
-	// OSIPhone::screen_width = rect.size.width - rect.origin.x;
-	// OSIPhone::screen_height = rect.size.height - rect.origin.y;
-
-	mainViewController = view_controller;
-
 	// prevent to stop music in another background app
 	[[AVAudioSession sharedInstance] setCategory:AVAudioSessionCategoryAmbient error:nil];
 
-	return TRUE;
+	return YES;
 };
 
 - (void)onAudioInterruption:(NSNotification *)notification {
-	if ([notification.name isEqualToString:AVAudioSessionInterruptionNotification]) {
-		if ([[notification.userInfo valueForKey:AVAudioSessionInterruptionTypeKey] isEqualToNumber:[NSNumber numberWithInt:AVAudioSessionInterruptionTypeBegan]]) {
-			NSLog(@"Audio interruption began");
-			on_focus_out(view_controller, &is_focus_out);
-		} else if ([[notification.userInfo valueForKey:AVAudioSessionInterruptionTypeKey] isEqualToNumber:[NSNumber numberWithInt:AVAudioSessionInterruptionTypeEnded]]) {
-			NSLog(@"Audio interruption ended");
-			on_focus_in(view_controller, &is_focus_out);
-		}
-	}
-};
+	NSLog(@"Audio interruption. Details: %@", notification.userInfo);
 
-- (void)applicationWillTerminate:(UIApplication *)application {
-	[self deinitGameControllers];
```

Comment:
Since the application will be jettisoned by iOS a few seconds after this callback, I would argue that it's not worth it to attempt to clean anything up here.

I don't feel very strongly about this either way however.

Reply:
Doesn't it help to save state / data before app terminates?